flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/StructuredToStringCastRule.java

```diff
@@ -0,0 +1,249 @@
+                boolean elementIsNull$301 = f3Value$295.isNullAt(i$299);
+                if (!elementIsNull$301) {
+                    element$300 = ((org.apache.flink.table.data.binary.BinaryStringData) f3Value$295.getString(i$299));
+                    builder$297.append(element$300);
+                } else {
+                    builder$297.append("NULL");
+                }
+            }
+            builder$297.append("]");
+            java.lang.String resultString$298;
+            resultString$298 = builder$297.toString();
+            result$7 = org.apache.flink.table.data.binary.BinaryStringData.fromString(resultString$298);
+            isNull$6 = result$7 == null;
+        } else {
+            result$7 = org.apache.flink.table.data.binary.BinaryStringData.EMPTY_UTF8;
+        }
+        builder$287.append("d=" + result$7);
+    } else {
+        builder$287.append("d=" + "NULL");
+    }
+    builder$287.append("}");
+    java.lang.String resultString$288;
+    resultString$288 = builder$287.toString();
+    result$1 = org.apache.flink.table.data.binary.BinaryStringData.fromString(resultString$288);
+
+     */
+    @Override
+    protected String generateCodeBlockInternal(
+            CodeGeneratorCastRule.Context context,
+            String inputTerm,
+            String returnVariable,
+            LogicalType inputLogicalType,
+            LogicalType targetLogicalType) {
+        StructuredType inputStructuredType = (StructuredType) inputLogicalType;
+
+        final String builderTerm = newName("builder");
+        context.declareClassField(
+                className(StringBuilder.class), builderTerm, constructorCall(StringBuilder.class));
+
+        final String resultStringTerm = newName("resultString");
+        final int length = LogicalTypeChecks.getLength(targetLogicalType);
+        final CastRuleUtils.CodeWriter writer =
+                new CastRuleUtils.CodeWriter()
+                        .stmt(methodCall(builderTerm, "setLength", 0))
+                        .stmt(methodCall(builderTerm, "append", strLiteral("{")));
+
+        for (int i = 0; i < inputStructuredType.getAttributes().size(); i++) {
+            final int fieldIndex = i;
+            final StructuredType.StructuredAttribute attribute =
+                    inputStructuredType.getAttributes().get(fieldIndex);
+
+            final String fieldTerm = newName("f" + fieldIndex + "Value");
+            final String fieldIsNullTerm = newName("f" + fieldIndex + "IsNull");
+
+            final CastCodeBlock codeBlock =
+                    CastRuleProvider.generateCodeBlock(
+                            context,
+                            fieldTerm,
+                            fieldIsNullTerm,
+                            // Null check is done at the row access level
+                            attribute.getType().copy(false),
```

Comment:
this could potentially break as e.g. `NullType` does not support `copy(false)`, unlikely but still

Reply:
You have a point here. This operation is performed in several rules. I will add an hotfix commit at the bottom of the history to fix it for all rules.